packages/@lwc/engine-core/src/patches/detect-synthetic-cross-root-aria.ts

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import {
+    assign,
+    globalThis,
+    isTrue,
+    KEY__NATIVE_GET_ELEMENT_BY_ID,
+    KEY__NATIVE_QUERY_SELECTOR_ALL,
+    isNull,
+    isUndefined,
+    getOwnPropertyDescriptor,
+    defineProperty,
+    ID_REFERENCING_ATTRIBUTES_SET,
+    isString,
+    isFunction,
+    StringSplit,
+    ArrayFilter,
+    hasOwnProperty,
+    KEY__SHADOW_TOKEN,
+} from '@lwc/shared';
+import { onReportingEnabled, report, ReportingEventId } from '../framework/reporting';
+import { getAssociatedVMIfPresent, VM } from '../framework/vm';
+import { logWarn } from '../shared/logger';
+
+//
+// The goal of this code is to detect invalid cross-root ARIA references in synthetic shadow DOM.
+// These invalid references should be fixed before the offending components can be migrated to native shadow DOM.
+// When invalid usage is detected, we warn in dev mode and call the reporting API if enabled.
+// See: https://lwc.dev/guide/accessibility#link-ids-and-aria-attributes-from-different-templates
+//
+
+// Use the unpatched native getElementById/querySelectorAll rather than the synthetic one
+const getElementById = globalThis[KEY__NATIVE_GET_ELEMENT_BY_ID] as typeof document.getElementById;
+
+const querySelectorAll = globalThis[
+    KEY__NATIVE_QUERY_SELECTOR_ALL
+] as typeof document.querySelectorAll;
+
+function isSyntheticShadowRootInstance(rootNode: Node): rootNode is ShadowRoot {
+    return rootNode !== document && isTrue((rootNode as any).synthetic);
+}
+
+function reportViolation(source: Element, target: Element, attrName: string) {
+    // The vm is either for the source, the target, or both. Either one or both must be using synthetic
+    // shadow for a violation to be detected.
+    let vm: VM | undefined = getAssociatedVMIfPresent((source.getRootNode() as ShadowRoot).host);
+    if (isUndefined(vm)) {
+        vm = getAssociatedVMIfPresent((target.getRootNode() as ShadowRoot).host);
+    }
+    if (isUndefined(vm)) {
+        // vm should never be undefined here, but just to be safe, bail out and don't report
+        return;
+    }
+    report(ReportingEventId.CrossRootAriaInSyntheticShadow, vm);
+    if (process.env.NODE_ENV !== 'production') {
+        logWarn(
+            `Element <${source.tagName.toLowerCase()}> uses attribute "${attrName}" to reference element ` +
```

Comment:
Food for thought. I have heard multiple times that developers have a hard time the origin of the warning. We do print the component stack today to make it easier. 

In that specific case, it might be also interesting to log the source and target elements in the warning log line. Hovering the elements in the log line would highlight the element in the DOM, this would also enable developers to right-click the element in the log line to reveal it in the "Element" panel.

I also heard in the past that objects logged in the console might not be GCed in certain browsers. This might have been fixed since then. 🤷‍♂️

Reply:
> In that specific case, it might be also interesting to log the source and target elements in the warning log line.

I considered doing that, but our current `logger.ts` and our tests would need to be modified to support that. You're right that it's a better DX.

> I also heard in the past that objects logged in the console might not be GCed in certain browsers. This might have been fixed since then. 🤷‍♂️

Ah yes, I forgot about this. No, [this is still a bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1370825) in Chrome at least.

OTOH it would only affect `PRODDEBUG` mode so I'm not sure if we care.

Reply:
Yeah I looked into it, and rearchitecting our entire logging system to support this would be a pretty heavy lift. We'd have to rewrite a bunch of the tests. I suggest we open an issue after this PR to address it.